flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java

```diff
@@ -0,0 +1,314 @@
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.io.WriteResult;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FilesCommitter implements Committer<FilesCommittable>, Serializable {
+  private static final long serialVersionUID = 1L;
+  private static final long INITIAL_CHECKPOINT_ID = -1L;
+  private static final Logger LOG = LoggerFactory.getLogger(FilesCommitter.class);
+  private static final String FLINK_JOB_ID = "flink.job-id";
+  private static final int MAX_RECOMMIT_TIMES = 3;
+
+  private static final String MAX_COMMITTED_CHECKPOINT_ID = "flink.max-committed-checkpoint-id";
+
+  // TableLoader to load iceberg table lazily.
+  private final TableLoader tableLoader;
+  private final boolean replacePartitions;
+  private final Map<String, String> snapshotProperties;
+
+  // It will have an unique identifier for one job.
+  private final Table table;
+  private transient Map<String, Long> maxCommittedCheckpointIdForJob = Maps.newHashMap();
+
+  private transient ExecutorService workerPool;
+
+  public FilesCommitter(
+      TableLoader tableLoader,
+      boolean replacePartitions,
+      Map<String, String> snapshotProperties,
+      int workerPoolSize) {
+    this.tableLoader = tableLoader;
+    this.replacePartitions = replacePartitions;
+    this.snapshotProperties = snapshotProperties;
+
+    this.workerPool =
+        ThreadPools.newWorkerPool(
+            "iceberg-worker-pool-" + Thread.currentThread().getName(), workerPoolSize);
+
+    // Open the table loader and load the table.
+    this.tableLoader.open();
+    this.table = tableLoader.loadTable();
+  }
+
+  @Override
+  public void commit(Collection<CommitRequest<FilesCommittable>> requests)
+      throws IOException, InterruptedException {
+    int dataFilesNumStored = 0;
+    int deleteFilesNumStored = 0;
+
+    Collection<CommitRequest<FilesCommittable>> store = Lists.newArrayList();
+    String jobId = null;
+    long maxCommittableCheckpointId = INITIAL_CHECKPOINT_ID;
+
+    for (CommitRequest<FilesCommittable> request : requests) {
+      WriteResult committable = request.getCommittable().committable();
+      Long committableCheckpointId = request.getCommittable().checkpointId();
+      jobId = request.getCommittable().jobID();
+      if (request.getNumberOfRetries() > MAX_RECOMMIT_TIMES) {
+        String message =
+            String.format(
+                "Failed to commit transaction %s after retrying %d times",
+                request.getCommittable(), MAX_RECOMMIT_TIMES);
+        request.signalFailedWithUnknownReason(new RuntimeException(message));
+      } else if (request.getNumberOfRetries() > 0) {
```

Comment:
Are we sure in this case that some parallel retries or anything has not committed this request to the table concurrently?

I have to check the other `if` branches too, but I think we always should check the current `maxCommittedCheckpointId` from the table.

Reply:
Do we want to try to commit a previously unsuccessful commit without considering the new data? Was this a conscious decision to do this separately? Wouldn't it be better to try to commit as few times as possible and try to commit this together with the other requests?

Reply:
Yes, I did deliberately separate the two.
Before I start to commit new data, I try to commit the data that needs to be retried in the last time. The main reason for doing this is that if I put it together with the current new data to be submitted, would it cause an unknowing error? I haven't thought of a use case like this yet. But if we are sure there is no problem, then this branch of judgment can be omitted.

Reply:
I would go with the simpler code first, and add new checks only if we find a specific case for it.

Reply:
Then I may need to change the data structure to a linked list, and put the data that restored from checkpoint to the head, let me try.